File: 4/solution.js
function largestPalindromeProduct(ceiling, floor) {
  var candidates = {}
      , counter = ceiling
      , product;

  while (counter > floor) {
    for (var i = ceiling; i > floor; i--) {
      product = i * counter;

      if (isPalindrome(product)) {
        candidates[i + ' * ' + counter] = product;
      }
    }
    counter--;
  }
  return findLargest(candidates);
}

function findLargest(candidates) {
  var largest = 0;

  for (var p in candidates) {
    if (candidates[p] > largest)
      largest = candidates[p];
  }
  return largest;
}

function isPalindrome(num) {
  var reverseNum = parseInt(num.toString().split('').reverse().join(''), 10);

  if (num === reverseNum) {
    return true;
  }
  else {
    return false;
  }
}

File: 4/solution.rb
def largest_palindrome_product(ceiling, floor)
  candidates, counter = {}, ceiling

  while counter > floor
    ceiling.downto(floor) do |i|
      product = i * counter

      if is_palindrome(product)
        candidates["#{i} * #{counter}"] = product
      end
    end
    counter -= 1
  end
  return find_largest(candidates)
end

def is_palindrome(num)
  reverse_num = num.to_s.reverse.to_i

  if num == reverse_num
    true
  else
    false
  end
end

def find_largest(candidates)
  largest = 0

  candidates.each do |key, value|
    if value > largest
      largest = value
    end
  end
  largest
end
